cmd/metal-api/internal/datastore/switch.go

```diff
@@ -196,3 +196,39 @@ func (rs *RethinkStore) ConnectMachineWithSwitches(m *metal.Machine) error {
 
 	return nil
 }
+
+// GetSwitchStatus get SwitchStatus for a given switch id
+func (rs *RethinkStore) GetSwitchStatus(id string) (*metal.SwitchStatus, error) {
+	rs.log.Infow("get state", "id", id)
+	var ss metal.SwitchStatus
+	err := rs.findEntityByID(rs.switchStatusTable(), &ss, id)
+	if err != nil && !metal.IsNotFound(err) {
+		return nil, err
+	}
+	if err != nil && metal.IsNotFound(err) {
+		return &metal.SwitchStatus{}, nil
+	}
+	rs.log.Infow("get state", "state", ss)
+
+	return &ss, nil
+}
+
+// SetSwitchStatus create or update the switch status.
+func (rs *RethinkStore) SetSwitchStatus(state *metal.SwitchStatus) (*metal.SwitchStatus, error) {
+	var ss metal.SwitchStatus
+	err := rs.findEntityByID(rs.switchStatusTable(), &ss, state.ID)
+	if err != nil && !metal.IsNotFound(err) {
+		return nil, err
+	}
+	if err != nil && metal.IsNotFound(err) {
+		rs.log.Infow("create state", "state", state)
+		createErr := rs.createEntity(rs.switchStatusTable(), state)
+		if err != nil {
+			return nil, createErr
+		}
+		return state, nil
+	}
+	rs.log.Infow("update state", "old", ss, "new", state)
+	err = rs.updateEntity(rs.switchStatusTable(), state, &ss)
+	return state, err
+}
```

Comment:
You can just use upsertEntity (see event insertions). Should reduce this function to a single line.

Reply:
But upsert does not create the entity if it does not exists, or do i miss something ?

Reply:
I think it does, it uses the same rethinkdb function as for creating entities, but does a replace on existing documents.

Reply:
done, and fixed the BUG in upsert not setting creation date